### ISSUE TITLE
[vcpkg-tool] `vcpkg install <pkg>` gives wrong usage information

### ISSUE THREAD

**Issue by dg0yt** [microsoft/vcpkg#20190]
# Bad User Experience

The auto-generated usage information is wrong for many non-trivial ports. 

## Earlier reports

 - #10881

## Known problem ports
 - `azure-iot-sdk-c` (lists 4 out of 10 targets in one line, too much)
 - `grpc` (references a `modules` target, #17177)
 - `hdf5` (listing all targets, which is too much)
 - `iir1`(config in wrong dir, https://github.com/microsoft/vcpkg/issues/23940)
 - `itk` (promoting bad practice, see below,  #22641)
 - `jsoncpp` (3 targets in one line, https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1082735779)
 - `libpng` (reporting different targets for static vs. dynamic instead of CMake Find module, #23379)
 - `openimageio` (references a `Module` target, #21234)
 - `openxr-loader` (config in wrong dir,  #23928)
 - `qhull` (listing all targets, which is too much)
 - `qt5-base` (listing targets which are private implementation details, and duplicate targets, cf. https://github.com/microsoft/vcpkg/issues/14907#issuecomment-1024871325-)
 - `realsense2` (lists four targets, https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1084750479)
 - `shapelib` (no config file to match the proposed package name, #16041, https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1124553735)

## Patched by usage file
 - `arrow` (references an `arrow_bundled_dependencies` library which doesn't exist in absence of vendored dependencies, #20459)
 - `c-ares` (lists `c-ares::cares c-ares::cares_shared c-ares::cares_static`)
 - `drogon`(picks targets from conditional `find_dependency` helpers, https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1008044822)
 - `libiconv` (see below, from #14994; #22539)
 - `mailio` (reports a `cmake` package, #23798)
 - `vtk` (original output cf. #19035)

## Other resolved

 - `effolkronium-random` (config in wrong dir, #20795)
 - `sentry-native` (config in wrong dir, https://github.com/microsoft/vcpkg/pull/24309#pullrequestreview-951333878)

## Misguided patching in vcpkg
 - `libgeotiff` 1.6.1 (#15750 diverging from upstream)

# Promoting Bad Practice

All targets are listed in the same `target_link_libraries` line. But normally, canonical usage is to use only a single desired target and let it pull in its transitive usage requirements. 

Some package do export targets not as public interface, but as an implementation detail hidden behind interface variables. By automatic reporting of targes, vcpkg distracts user from applying canonical usage, contributors from researching, documenting and testing canonical usage, and team members to base there support answers on canonical usage.

# Example: User experience for libiconv

> but when i run the command  `vcpkg install libiconv`, it always tell me
> ```
> ubuntu@o-01:~/vcpkg$ vcpkg install libiconv
> Computing installation plan...
> The following packages are already installed:
>     libiconv[core]:x64-linux -> 1.16#11
> Package libiconv:x64-linux is already installed
> Restored 0 packages from /home/ubuntu/.cache/vcpkg/archives in 1.052 us. Use --debug to see more details.
> 
> Total elapsed time: 37.06 us
> 
> The package libiconv provides CMake targets:
> 
>     find_package(iconv CONFIG REQUIRED)
>     target_link_libraries(main PRIVATE Iconv::Charset)
> ```
> 
> please fix the issue.

_Originally posted by @ssrlive in https://github.com/microsoft/vcpkg/issues/14994#issuecomment-920596707_

**Comment by dg0yt** [comment]
The `CONFIG` keyword is wrong. 

Generally, I think that given a `share/<pkg>/vcpkg-cmake-wrapper.cmake`, the tool heuristics would need to look for a `Find<pkg>.cmake` (case insensitively), and suggest `find_package(<pkg-with-rectified-case> REQUIRED)` if such a module is found.

**Comment by dg0yt** [comment]
In addition, `Iconv::Charset` is an internal helper of the wrapper, not a public feature.

**Comment by dg0yt** [comment]
Is there sufficient evidence now  to request a response from the vcpkg team?

**Comment by dg0yt** [comment]
CC @ras0219-msft @strega-nil-ms

**Comment by strega-nil-ms** [comment]
This is fair enough; I'll bring it up in the new year at the PR meeting.

**Comment by dg0yt** [comment]
~~~
The following packages will be built and installed:
    drogon[core,mysql,orm]:x64-linux -> 1.7.4
....
The package drogon provides CMake targets:

    find_package(Drogon CONFIG REQUIRED)
    # Note: 6 target(s) were omitted.
    target_link_libraries(main PRIVATE pg_lib UUID_lib coz::coz MySQL_lib)
~~~
The Drogon config clearly provides target `Drogon::Drogon`. None of the other targets should be linked manually. They seem to come from Find modules installed but only conditionally used via `find_dependency(...)`. Maybe a hint how to improve the heuristic.

**Comment by dg0yt** [comment]
This is a friendly reminder. Actually I'm increasingly unhappy with how this vcpkg feature promotes bad practice.

**Comment by dg0yt** [comment]
Added `qt5-base`.

**Comment by dg0yt** [comment]
Added `azure-iot-sdk-c`

**Comment by dg0yt** [comment]
Added `qhull`.

Do I need to make a PR which disables the tool feature?

**Comment by dg0yt** [comment]
Added `libpng`.

Ping @vicroms @strega-nil-ms.

**Comment by dg0yt** [comment]
Today's highlight from my draft gdal-cmake build, after continuing work on exported CMake config.

~~~
The package gdal provides CMake targets:

    find_package(3.12 CONFIG REQUIRED)
    # Note: 6 target(s) were omitted.
    target_link_libraries(main PRIVATE PNG::PNG JPEG::JPEG Boost::boost EXPAT::EXPAT)

    find_package(3.13 CONFIG REQUIRED)
    target_link_libraries(main PRIVATE XercesC::XercesC)

    find_package(3.14 CONFIG REQUIRED)
    target_link_libraries(main PRIVATE CURL::libcurl)

    find_package(3.16 CONFIG REQUIRED)
    target_link_libraries(main PRIVATE PostgreSQL::PostgreSQL)

    find_package(GDAL CONFIG REQUIRED)
    target_link_libraries(main PRIVATE GDAL::GDAL)

    find_package(packages CONFIG REQUIRED)
    # Note: 59 target(s) were omitted.
    target_link_libraries(main PRIVATE DAP::DAP ECW::ECW GIF::GIF IDB::IDB)
~~~

As already mentioned [two months ago](https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1008044822), it is wrong to blindly report all find modules which are used by a cmake config.

**Comment by dg0yt** [comment]
@vicroms @strega-nil-ms @ras0219-msft @BillyONeal Issue open for 6 months with no adequate reaction. I could contribute to a solution more actively, but I refuse to work on these grounds.

**Comment by JackBoosY** [comment]
We will discuss this in the next meeting.
Sorry for late reply.

**Comment by BillyONeal** [comment]
> @vicroms @strega-nil-ms @ras0219-msft @BillyONeal Issue open for 6 months with no adequate reaction. I could contribute to a solution more actively, but I refuse to work on these grounds.

I think you're not getting a response here because the problem described is really a structural problem that's going to require massive investment / spec / testing / etc. to resolve, and while the experience can be better, this bug hasn't risen to the top of priorities to spend multiple developer weeks to land a solution.

@dan-shaw @ras0219 @strega-nil @vicroms and I just discussed this in person with potential things we could do being encouraging people to provide more usage files when we merge things into the repo, or we could provide entries in vcpkg.json to make it easier to generate better usage information that we could test. We have been hesitant to add that until we had the time/capability to do checking / enforcement that the resulting find_package calls would work.

On the other hand we aren't sure how to interpret your message from 3 days ago ( https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1067609076 ); it seems like nothing about how we maintain the catalog could make you happy here and nothing short of removal of all generated usages. On the other hand that may be a specific bug we could improve, but since this is stapled into the overall long term structural issue we aren't sure how to interpret your message.

**Comment by dg0yt** [comment]
I'm not sure if it is better to have users open the same issues again and again.
Collection the issues was meant to help to incrementally approach low hanging fruits.

- Implement a strict review policy.
  - Don't accept new ports without verified usage (implicit or explicit).
  - Don't accept non-security updates which leave broken usage unfixed (example: #23510).
- Never list more than one target in `target_link_libraries(main PRIVATE ...)`. 
  (We can assume that targets properly carry transitive usage requirements unless upstream is broken.)
  Test cases: azure-iot-sdk-c, drogon, civetweb
- Ignore all Find modules. 
  (We can assume they are just helpers or backports).
  Test cases: civetweb (more candidates via CI file lists) 
- Find module usage:
  If there is `vcpkg-cmake-wrapper.cmake`, look for a `<Pkg>::<Pkg>` target, `<Pkg>_INCLUDE_DIRS` and `<Pkg>_LIBRARIES` in the official Find module from CMake. 
  Maybe even print a link to the find module documentation url.
  Don't take variables or targets from `vcpkg-cmake-wrapper`.
  Test case: libiconv / FindIconv.cmake.
- Config file usage:
  If there is a `<Pkg>Config.cmake` or `<pkg>-config.cmake` in a suitable directory, find `<basename>-release.cmake` and pick targets from `<basename>.cmake`. Prioritize by name `<pkg>::lib<pkg>`, `<pkg>::<pkg>, or by link libraries relationships.
  Test cases: drogon, civetweb
- Basically, if there is still more than one target, don't report any target. 
  (Issue a mesage to remind maintainers/reviewers to add a usage file.)
- (And then there is pkgconfig usage, even with targets now in CMake.)

**Comment by dg0yt** [comment]
> Implement a strict review policy.
> * Don't accept new ports without verified usage (implicit or explicit).
> * Don't accept non-security updates which leave broken usage unfixed (example: [[libiconv] remove doc files from share #23510](https://github.com/microsoft/vcpkg/pull/23510)).

@BillyONeal Recent example for a port which is already listed here, with a PR just approved while facing unusable usage information: https://github.com/microsoft/vcpkg/pull/23502#issuecomment-1073655939

**Comment by dg0yt** [comment]
Added `mailio`.

**Comment by dg0yt** [comment]
Adding `jsoncpp`. From https://github.com/microsoft/vcpkg/issues/23839#issuecomment-1082709408.
~~~
The package jsoncpp provides CMake targets:

    find_package(jsoncpp CONFIG REQUIRED)
    target_link_libraries(main PRIVATE jsoncpp_object jsoncpp_static JsonCpp::JsonCpp)
~~~

**Comment by JackBoosY** [comment]
We really need a post check to check the usages.

**Comment by dg0yt** [comment]
> We really need a post check to check the usages.

Yes. It is wasting so much time to check, answer and repair. Again and again.

**Comment by dg0yt** [comment]
> We really need a post check to check the usages.

However, the check can't detect all pointless proposal. It won't detect overlinking.

**Comment by dg0yt** [comment]
Adding `realsense2`. Output:
~~~
   target_link_libraries(main PRIVATE realsense2::fw realsense2::usb realsense2::realsense2 realsense2::realsense-file)
~~~
Reasonable:
~~~
   target_link_libraries(main PRIVATE realsense2::realsense2)
   target_link_libraries(main PRIVATE realsense2::fw)
~~~

I'm tempted to call this the "realnonsense" bug from now.

**Comment by dg0yt** [comment]
Adding `iir1` and `openxr-loader`: Config in wrong dir. This could be noticed by the post-build check...

**Comment by dg0yt** [comment]
A particular issue is printing directory names instead of module names. This is what leads to 
~~~
find_package(cmake ...) # share/<port>/cmake/....cmake, e.g. mailio
find_package(3.14 ...) # share/<port>/3.14/Find<Pkg>.cmake, e.g. gdal

**Comment by JackBoosY** [comment]
> A particular issue is printing directory names instead of module names. This is what leads to
> 
> ```
> find_package(cmake ...) # share/<port>/cmake/....cmake, e.g. mailio
> find_package(3.14 ...) # share/<port>/3.14/Find<Pkg>.cmake, e.g. gdal
> ```

Normally. it should be caused by the incorrect installation path.

**Comment by dg0yt** [comment]
> Normally. it should be caused by the incorrect installation path.

For mailio, perhaps. But then vcpkg tool would need to report a validation error, not an unusable usage information.
For GDAL, no. These are utility find modules, needed for dependencies of the exported config (which is in the correct directory). 

Basically, vcpkg tool must ignore for usage what isn't found by our CMake toolchain. Our CMake toolchain finds config or wrappers only directly in `share/<pkg>`. This is the base for the proposal in https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1072067545.

**Comment by strega-nil-ms** [comment]
@dg0yt, at our meeting, we discussed this, and these are our thoughts:

> I'm not sure if it is better to have users open the same issues again and again. Collecting the issues was meant to help adjust priorities and to incrementally approach low hanging fruits.
> 
> * Implement a strict review policy.
>   
>   * Don't accept new ports without verified usage (implicit or explicit).
>   * Don't accept non-security updates which leave broken usage unfixed (example: [[libiconv] remove doc files from share #23510](https://github.com/microsoft/vcpkg/pull/23510)).

Short-term, this would be _really_ difficult to do in a way that's not prohibitively expensive on our engineers' time.
Long-term, this would be really nice, tho.

We could add specific requirements to the maintainer guidelines for new ports, to review the autogenerated usage and make sure the `find_package` call resolves.

> * Never list more than one target in `target_link_libraries(main PRIVATE ...)`.
>   (We can assume that targets properly carry transitive usage requirements unless upstream is broken.)
>   Test cases: azure-iot-sdk-c, drogon, civetweb
> * Basically, if there is still more than one target, don't report any target.
>   (Issue a mesage to remind maintainers/reviewers to add a usage file.)

(later bullet moved up) We would like to continue listing more than one target, but perhaps with a warning message about adding a usage file, and warn/fail in CI

> * Ignore all Find modules.
>   (We can assume they are just helpers or backports).
>   Test cases: civetweb (more candidates via CI file lists)

I think we try to do this already, but we definitely should do this more if we ever don't

> * Find module usage:
>   If there is `vcpkg-cmake-wrapper.cmake`, look for a `<Pkg>::<Pkg>` target, `<Pkg>_INCLUDE_DIRS` and `<Pkg>_LIBRARIES` in the official Find module from CMake.
>   Maybe even print a link to the find module documentation url.
>   Don't take variables or targets from `vcpkg-cmake-wrapper`.
>   Test case: libiconv / FindIconv.cmake.

Agree that this should be a usage file, but doing this automatically would be difficult since we don't have a mapping of modules <-> ports

> * Config file usage:
>   If there is a `<Pkg>Config.cmake` or `<pkg>-config.cmake` in a suitable directory, find `<basename>-release.cmake` and pick targets from `<basename>.cmake`. Prioritize by name `<pkg>::lib<pkg>`, `::, or by link libraries relationships.
>   Test cases: drogon, civetweb

We think this should be subsumed by "if we detect multiple targets, tell the user to add a usage file" - we don't want to be heuristicking like this.

> * (And then there is pkgconfig usage, even with targets now in CMake.)

It would be good to add this feature to vcpkg (with a similar check for "if we have multiple pc files, add a usage")

On the other hand, this would likely be quite expensive, and the vcpkg team is unlikely to be able to do this in the short-term (or even medium-term);
especially making sure that the CI fails would be a ton of work (see @autoantwort's work on absolute paths as an example of this).
However, a useful minor change we could take from this is, for example, making our heuristically generated usage information less authoritative;
perhaps adding a `# this is heuristically generated, and may not be correct` to the output would be good.

**Comment by dg0yt** [comment]
> However, a useful minor change we could take from this is, for example, making our heuristically generated usage information less authoritative;
> perhaps adding a `# this is heuristically generated, and may not be correct` to the output would be good.

This sounds like a good first step.

**Comment by dg0yt** [comment]
Recent additions: c-ares, sentry-native

**Comment by dg0yt** [comment]
Added `shapelib`:
~~~
shapelib provides CMake targets:
    # this is heuristically generated, and may not be correct
    find_package(shp CONFIG REQUIRED)
    target_link_libraries(main PRIVATE shp)
~~~

~~~
$ ls packages/shapelib_x64-linux/share/shp/
export_shp.cmake  export_shp-debug.cmake  export_shp-release.cmake
~~~

**Comment by dg0yt** [comment]
`shapelib` bonus fact:
#22633 added a patch file to fix the config, but didn't add it to the portfile.
Not detected in review.
Not detected in automated checks.
And IMO the patch is invalid because it makes assumptions about the desired package name. It must be unofficial until fixed upstream.

**Comment by dg0yt** [comment]
LOL while working on #24720:
~~~
Elapsed time to handle osgearth:x64-linux: 16.23 min

Total elapsed time: 16.28 min

osgearth is header-only and can be used from CMake via:

    find_path(OSGEARTH_INCLUDE_DIRS "osgEarth/AGG.h")
    target_include_directories(main PRIVATE ${OSGEARTH_INCLUDE_DIRS})
~~~
(Libs went to `/lib64` by error. Headers don't have a `.h` suffix except for the one listed.)

Where is the `# this is heuristically generated, and may not be correct`?

**Comment by dg0yt** [comment]
The tool also creates invalid suggestions by copying things literally, e.g.
`target_link_libraries(main PRIVATE <name> alpaka alpaka::alpaka)` (https://github.com/microsoft/vcpkg/pull/26812#issuecomment-1250955731), or
`target_link_libraries(main PRIVATE # foobar)`

**Comment by dg0yt** [comment]
> * From this thread, I understand that the current recommended way is to add a usage file in the port folder ?
 
yes.

> * If so, is there any specific template or format to follow for getting quick approval of the PR ?

Look at a similar port, or start with the tools ouput. But use your brain, and upstream documentation.

> I remember getting quite a few times # note: X additional targets are not displayed. forcing me to dig the corresponding share/<lib> folder.

Quite often, there are only a few relevant targets, and the others are used implicitly.

**Comment by JoseSanchez7** [comment]
Hello @dg0yt. In regard to the azure-iot-sdk-c, what exactly are you referring to? An explicit example in the code base would be helpful.

**Comment by dg0yt** [comment]
@JoseSanchez7 Without looking at the current state:

> azure-iot-sdk-c (lists 4 out of 10 targets in one line, too much)

vcpkg detects many targets and print some 4 of them. This is not a good path to indicate usage. In general,
- most targets are for a particular purpose. They shouldn't be used when they aren't needed.
- some targets are are common code.They shouldn't be used directly, but are included as transitive usage requirements.

The proper fix is to add a `usage` file to the port, connecting targets to purposes. Simple example: `ports/geos/usage`.
Or just the most common case, and a pointer to upstream documentation. (Unfortunately, many upstream are particular in this aspect of documentation.)

**Comment by JoseSanchez7** [comment]
Hello @dg0yt, our team (Azure C SDK) spoke yesterday morning about your request regarding changes to how our cmake works with vcpkg and decided that it's not something we will be changing at this moment.

**Comment by dg0yt** [comment]
@JoseSanchez7 I didn't want to indicate a need for changes to "how your cmake works". I see a need to better document downstream CMake usage. The heuristics in vcpkg tool is very limited. It is unlikely that a random pick of 4 targets out of 10 is what the users needs.

**Comment by dg0yt** [comment]
@BillyONeal Not sure if it was case the before the last tool release, but now the heuristics picks misleading targets from the wrappers:
~~~
tiff provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(tiff CONFIG REQUIRED)
    target_link_libraries(main PRIVATE ZSTD::ZSTD)
~~~

**Comment by dg0yt** [comment]
@BillyONeal Ignore my last comment. I bootstrapped on the old version. The new version no longer outputs misleading information for tiff.
(Now we should probably looka at my next proposal: Check if there is an official find module when we find wrapper.)

**Comment by Cheney-W** [comment]
Stay active
